File: leetcode/分类/动态规划/1373. 二叉搜索子树的最大键值和.cpp
#include <algorithm>
#include <queue>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

using namespace std;

class Solution {
public:
    typedef struct info {
        bool isBST;
        int sum;
        int left;
        int right;
    } info;

    int ans = 0;

    int maxSumBST(TreeNode *root) {
        dfs(root);
        return ans;
    }

    info dfs(TreeNode *root) {
        if (!root) {
            info i;
            i.isBST = true;
            i.left = INT_MAX;
            i.right = INT_MIN;
            i.sum = 0;
            return i;
        }

        auto left = dfs(root->left);
        auto right = dfs(root->right);

        info currInfo;
        currInfo.isBST = true;
        int val = root->val;
        if (!left.isBST || !right.isBST || val <= left.right || val >= right.left) {
            currInfo.isBST = false;
            currInfo.left = INT_MIN;
            currInfo.right = INT_MAX;
            currInfo.sum = -1;
            return currInfo;
        }

        currInfo.sum = val + left.sum + right.sum;
        ans = max(ans, currInfo.sum);
        currInfo.left = min(left.left, val);
        currInfo.right = max(val, right.right);
        return currInfo;
    }
};

File: leetcode/1300-1400/1373. 二叉搜索子树的最大键值和.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from operator import *
from typing import *
from queue import *
from string import *
from random import *

DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]
Max = lambda a, b: b if b > a else a
Min = lambda a, b: b if b < a else a
INF = float("inf")
MOD = int(1e9 + 7)


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxSumBST(self, root: Optional[TreeNode]) -> int:
        ans = 0

        def dfs(node: Optional[TreeNode]):
            if not node:
                return (inf, -inf, 0)

            t = node.val

            l_mn, l_mx, l = dfs(node.left)
            r_mn, r_mx, r = dfs(node.right)

            if l_mx >= t or r_mn <= t:
                return (-inf, inf, 0)
            nonlocal ans
            ans = Max(ans, l + r + t)

            return (min(l_mn, r_mn, t), max(l_mx, r_mx, t), l + r + t)

        dfs(root)
        return ans
